File: platform/src/main/java/com/auraboot/framework/meta/validation/RuleMerger.java
package com.auraboot.framework.meta.validation;

import com.auraboot.framework.meta.dto.CrossFieldRule;
import com.auraboot.framework.meta.dto.RuleOverride;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Merges model-level rules with command-level overrides.
 *
 * Merge strategy:
 * - override with disabled=true → remove rule
 * - override with matching id → replace rule entirely
 * - override with new id → append as new rule
 *
 * disabled takes precedence over assert (if both set, rule is removed).
 */
public final class RuleMerger {

    private RuleMerger() {}

    public static List<CrossFieldRule> merge(
            List<CrossFieldRule> modelRules,
            List<RuleOverride> overrides) {

        if (overrides == null || overrides.isEmpty()) {
            return new ArrayList<>(modelRules);
        }

        // Build override map preserving insertion order
        Map<String, RuleOverride> overrideMap = new LinkedHashMap<>();
        for (RuleOverride o : overrides) {
            overrideMap.put(o.getId(), o);
        }

        List<CrossFieldRule> result = new ArrayList<>();

        // Process model rules: apply matching overrides
        for (CrossFieldRule rule : modelRules) {
            RuleOverride override = overrideMap.remove(rule.getId());
            if (override == null) {
                // No override — keep original
                result.add(rule);
            } else if (Boolean.TRUE.equals(override.getDisabled())) {
                // Disabled — skip (remove rule)
            } else {
                // Replace with override
                result.add(override);
            }
        }

        // Append remaining overrides (new rules not in model)
        for (RuleOverride override : overrideMap.values()) {
            if (!Boolean.TRUE.equals(override.getDisabled())) {
                result.add(override);
            }
        }

        return result;
    }
}

File: web-admin/app/framework/meta/validation/ruleMerger.ts
import type { CrossFieldRule, RuleOverride } from './crossFieldRuleTypes';

/**
 * Merge model-level rules with command-level overrides.
 * - disabled=true → remove rule
 * - matching id → replace
 * - new id → append
 */
export function mergeRules(
  modelRules: CrossFieldRule[],
  overrides: RuleOverride[],
): CrossFieldRule[] {
  if (!overrides || overrides.length === 0) {
    return [...modelRules];
  }

  const overrideMap = new Map<string, RuleOverride>();
  for (const o of overrides) {
    overrideMap.set(o.id, o);
  }

  const result: CrossFieldRule[] = [];

  for (const rule of modelRules) {
    const override = overrideMap.get(rule.id);
    if (override) {
      overrideMap.delete(rule.id);
      if (override.disabled) continue; // Remove
      result.push(override); // Replace
    } else {
      result.push(rule);
    }
  }

  // Append new overrides
  overrideMap.forEach((override) => {
    if (!override.disabled) {
      result.push(override);
    }
  });

  return result;
}
